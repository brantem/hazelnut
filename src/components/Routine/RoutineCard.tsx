import { EllipsisVerticalIcon } from '@heroicons/react/20/solid';

import type { Routine as _Routine } from 'types/routine';

type RoutineProps = {
  routine: _Routine;

  onSettingsClick: () => void;
};

const Routine = ({ routine, onSettingsClick }: RoutineProps) => {
  return (
    <div className={`px-4 py-3 bg-${routine.color}-50`} data-testid="routine-card">
      <div className="flex justify-between items-center space-x-3">
        <h3 className={`uppercase text-sm font-semibold text-${routine.color}-600 truncate`}>{routine.title}</h3>

        <div className="flex items-center space-x-1 flex-shrink-0">
          <button
            className={`p-1 rounded-md hover:bg-${routine.color}-100`}
            onClick={onSettingsClick}
            data-testid="routine-card-settings"
          >
            <EllipsisVerticalIcon className="h-5 w-5" />
          </button>
        </div>
      </div>
    </div>
  );
};

export default Routine;
