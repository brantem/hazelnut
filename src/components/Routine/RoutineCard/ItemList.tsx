import { useState, useCallback } from 'react';
import {
  DndContext,
  closestCenter,
  MouseSensor,
  TouchSensor,
  KeyboardSensor,
  useSensor,
  useSensors,
  UniqueIdentifier,
} from '@dnd-kit/core';
import {
  arrayMove,
  SortableContext,
  useSortable,
  sortableKeyboardCoordinates,
  verticalListSortingStrategy,
} from '@dnd-kit/sortable';
import { restrictToVerticalAxis, restrictToParentElement } from '@dnd-kit/modifiers';
import { CSS } from '@dnd-kit/utilities';
import clsx from 'clsx';

import Checkbox from 'components/Checkbox';

import { useItemsStore, useRoutinesStore, useHistoriesStore } from 'lib/stores';
import { Routine } from 'types/routine';
import { Item as _Item } from 'types/item';

type ItemProps = {
  routine: Routine;
  item: _Item;
  isSortable?: boolean;
};

const Item = ({ routine, item, isSortable }: ItemProps) => {
  const isDone = useHistoriesStore(
    useCallback((state) => state.getIsDone(routine.id, item.id, true), [routine.id, item.id]),
  );
  const save = useHistoriesStore((state) => state.save);
  const { attributes, listeners, setNodeRef, setActivatorNodeRef, transform, transition } = useSortable({
    id: item.id,
  });

  return (
    <li
      ref={setNodeRef}
      className="flex h-8 w-full items-center justify-between space-x-2 pr-1"
      style={{ transform: CSS.Transform.toString(transform), transition }}
      {...attributes}
    >
      {isSortable && (
        <button
          className="-ml-1.5 flex-grow-0 p-1.5"
          ref={setActivatorNodeRef}
          {...listeners}
          data-testid="routine-item-handle"
        >
          <svg viewBox="0 0 20 20" width="12" className="h-4 w-4 text-neutral-500" fill="currentColor">
            <path d="M7 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 2zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 14zm6-8a2 2 0 1 0-.001-4.001A2 2 0 0 0 13 6zm0 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 14z"></path>
          </svg>
        </button>
      )}

      <div className={clsx('w-full', isSortable && 'max-w-[calc(100%-theme(spacing.8))]')}>
        <Checkbox
          label={item.title}
          name={routine.id + '-' + item.id}
          color={routine.color}
          checked={isDone}
          onChange={(e) => save(routine, item, e.target.checked)}
        />
      </div>
    </li>
  );
};

type ItemListProps = {
  routine: Routine;
  isSortable?: boolean;
};

const ItemList = ({ routine, isSortable }: ItemListProps) => {
  const edit = useRoutinesStore((state) => state.edit);
  const items = useItemsStore(useCallback((state) => state.getItemsByIds(routine.itemIds), [routine.itemIds]));
  const sensors = useSensors(
    useSensor(MouseSensor),
    useSensor(TouchSensor),
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates }),
  );

  const [itemId, setItemId] = useState<UniqueIdentifier | null>(null);
  const activeIndex = items.findIndex((item) => item.id === itemId);

  if (!items.length) return null;

  return (
    <DndContext
      sensors={sensors}
      collisionDetection={closestCenter}
      onDragStart={({ active }) => active && setItemId(active.id)}
      onDragEnd={({ over }) => {
        setItemId(null);
        if (!over) return;
        const overIndex = items.findIndex((item) => item.id === over.id);
        if (activeIndex === overIndex) return;
        edit(routine.id, { itemIds: arrayMove(items.map((item) => item.id), activeIndex, overIndex) }); // prettier-ignore
      }}
      onDragCancel={() => setItemId(null)}
      modifiers={[restrictToVerticalAxis, restrictToParentElement]}
    >
      <SortableContext items={items} strategy={verticalListSortingStrategy}>
        <ol className="space-y-1 pt-1" data-testid="routine-card-items">
          {items.map((item) => (
            <Item key={item.id} routine={routine} item={item} isSortable={isSortable} />
          ))}
        </ol>
      </SortableContext>
    </DndContext>
  );
};

export default ItemList;
