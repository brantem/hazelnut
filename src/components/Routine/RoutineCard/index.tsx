import { EllipsisHorizontalIcon, ChevronUpIcon } from '@heroicons/react/20/solid';
import clsx from 'clsx';

import Card from 'components/Card';
import ItemList from 'components/Routine/RoutineCard/ItemList';

import { useRoutinesStore } from 'lib/stores';
import type { Routine } from 'types/routine';
import * as constants from 'data/constants';
import { useModal } from 'lib/hooks';

type RoutineProps = {
  routine: Routine;
  showAction?: boolean;
  isItemSortable?: boolean;
};

const RoutineCard = ({ routine, showAction, isItemSortable = false }: RoutineProps) => {
  const { setRoutine, edit } = useRoutinesStore((state) => ({ setRoutine: state.setRoutine, edit: state.edit }));
  const saveItemsModal = useModal(constants.modals.saveItemsToRoutine);
  const settingsModal = useModal(constants.modals.routineSettings);

  return (
    <Card
      color={routine.color}
      title={
        <>
          <span>{routine.title}</span>
          {routine.time && (
            <span className={`text-sm font-medium text-${routine.color}-400 ml-2 flex-shrink-0 tabular-nums`}>
              {routine.time}
            </span>
          )}
        </>
      }
      actions={[
        {
          children: 'Items',
          onClick: () => {
            setRoutine(routine);
            saveItemsModal.show();
          },
          testId: 'routine-card-save-items',
          skip: !showAction,
        },
        {
          children: <EllipsisHorizontalIcon className="h-5 w-5" />,
          onClick: () => {
            setRoutine(routine);
            settingsModal.show();
          },
          testId: 'routine-card-settings',
          skip: !showAction,
        },
        {
          children: <ChevronUpIcon className={clsx('h-5 w-5', routine.minimized && 'rotate-180')} />,
          onClick: () => edit(routine.id, { minimized: !routine.minimized }),
          testId: 'routine-card-minimize',
        },
      ]}
      data-testid="routine-card"
    >
      {!routine.minimized && <ItemList routine={routine} isSortable={isItemSortable} />}
    </Card>
  );
};

export default RoutineCard;
