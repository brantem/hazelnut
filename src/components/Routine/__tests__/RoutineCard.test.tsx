import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';

import RoutineCard from 'components/Routine/RoutineCard';

test('RoutineCard', () => {
  const onSettingsClick = vi.fn(() => {});
  const { container } = render(
    <RoutineCard routine={{ id: 'routine-1', title: 'Routine 1', color: 'red' }} onSettingsClick={onSettingsClick} />,
  );

  expect(container).toMatchSnapshot();

  act(() => screen.getByTestId('routine-card-settings').click());
  expect(onSettingsClick).toHaveBeenCalled();
});
