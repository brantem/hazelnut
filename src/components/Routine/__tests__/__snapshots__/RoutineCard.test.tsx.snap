// Vitest Snapshot v1

exports[`RoutineCard 1`] = `
<div>
  <div
    class="px-4 py-3 bg-red-50"
    data-testid="routine-card"
  >
    <div
      class="flex justify-between items-center space-x-3"
    >
      <h3
        class="uppercase text-sm font-semibold text-red-600 truncate"
      >
        Routine 1
      </h3>
      <div
        class="flex items-center space-x-1 flex-shrink-0"
      >
        <button
          class="p-1 rounded-md hover:bg-red-100"
          data-testid="routine-card-settings"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 3a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM10 8.5a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM11.5 15.5a1.5 1.5 0 10-3 0 1.5 1.5 0 003 0z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
